### PR TITLE
Gpu transform

### PR DESCRIPTION
**Pull request by slayton58**
Adds a thread pool for image decode, and optional GPU-based data conversion, mean subtraction and std division

### CODE REVIEW

File: caffe2/image/image_input_op.h [hunk 1]
```diff
@@ -192,80 +364,55 @@ bool ImageInputOp<Context>::Prefetch() {
...
   }
-  #pragma omp parallel for num_threads(4)
+
   for (int item_id = 0; item_id < batch_size_; ++item_id) {
     std::bernoulli_distribution mirror_this_image(0.5);
-    std::mt19937& randgen = randgen_per_thread[omp_get_thread_num()];
-    float* image_data = prefetched_image_.mutable_data<float>()
-        + crop_ * crop_ * channels * item_id;
-    string key, value;
+    std::mt19937& randgen = randgen_per_thread[num_decode_threads_];
```
Comment (akyrola) [inline caffe2/image/image_input_op.h:383]: I think here is a bug, since randgen_per_thread has only num_decode_threads_ elements. My code crashes here.

**Comment by slayton58** [comment]
Damnit, we had a race between when the PR was imported and when I found + fixed that issue: https://github.com/slayton58/caffe2/commit/50a8505a6a89a492a96900c54e505245e3bc03f8

File: caffe2/image/transform_gpu.cu [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  }
+}
+
+}
+
+template <typename T_IN, typename T_OUT, class Context>
+bool TransformOnGPU(Tensor<Context>& X, Tensor<Context> *Y, T_OUT mean, T_OUT std, Context *context) {};
+
+template <>
+bool TransformOnGPU<uint8_t, float, CUDAContext>(Tensor<CUDAContext>& X, Tensor<CUDAContext> *Y, float std, float mean, CUDAContext *context)
```
Comment (Yangqing) [inline caffe2/image/transform_gpu.cu:49]: @slayton58 Simon, @dutran caught inconsistency - seems that we are having std and mean instead of the other order round, is this intended? Or is it a bug actually?

Comment (slayton58) [reply caffe2/image/transform_gpu.cu:49]: This was a bug, and was fixed previously (see my comment on 01/09 in fact!)

It was fixed by me post-importing, but pre-merging -- I think that's a corner case for your import routines

Comment (Yangqing) [reply caffe2/image/transform_gpu.cu:49]: T_T

Comment (slayton58) [reply caffe2/image/transform_gpu.cu:49]: This reminds me, need to add a test for this code...
